WordPress/wordpress-develop
# Ensure that first content image is not lazy-loaded in block themes either

* This PR addresses all specific problems that causes the lazy-loading logic to not work correctly in block themes, covering **both content images and featured images** (i.e. `post-content` block and `post-featured-image` block).
* In general, for either of them, one problem is that `wp_filter_content_tags()` runs over the entire template content after both the featured images and the post content images have already been processed once though. While the underlying problem for that is fixed in #3549, this PR fixes the `loading="lazy"` problem specifically in a decoupled manner, ensuring that by default the function always returns `false` when parsing the entire block template (via a new `the_block_template` context).
* With that, there is still a remaining problem though that specifically the featured image in the `post-featured-image` block would still receive `loading="lazy"` by default. The reason for that is that, in a block theme, the relevant logic runs outside of the loop - which strictly speaking is a backward-compatibility break in the first place. A fix for the exact same problem is already present in the `post-content` block, so this PR simply adds the same fix to the `post-featured-image` block.
* With these changes, the problems are addressed. Tests are included to verify all of the fixes work correctly.

Trac ticket: https://core.trac.wordpress.org/ticket/56930

### Test instructions

Any images can be used for the below tests.

1. Use a block theme, e.g. Twenty Twenty-Three, with its factory templates.
2. **Test 1:** Create a post with 2 `core/image` blocks. When viewing them in the frontend, ensure that the first image does not have a `loading="lazy"` attribute, while the second _does_ have it.
3. **Test 2:** On the same post, set a featured image. When viewing the home page (which includes that post and its featured image), ensure that the featured image does not have a `loading="lazy"` attribute.
4. **Test 3:** Go to the Site Editor and edit the "Header" template part. Insert a `core/image` block anywhere in it. Then do the same for the "Footer" template part, insert a `core/image` block anywhere in it. When viewing the frontend, ensure that the image you placed in the "Header" does not have a `loading="lazy"` attribute, while the image you placed in the "Footer" _does_ have it.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
